internal/app/daemon/daemon.go

```diff
@@ -0,0 +1,321 @@
+
+	// Wait for Logstash and GRPC Server to shutdown
+	serverStopComplete := false
+	logstashInstanceStopComplete := false
+	for !serverStopComplete || !logstashInstanceStopComplete {
+		select {
+		case <-t.C:
+			d.log.Debug("Shutdown timeout reached, force shutdown.")
+			d.server.Stop()
+			serverStopComplete = true
+			logstashInstanceStopComplete = true
+		case <-serverStopped:
+			d.log.Debug("server successfully stopped.")
+			serverStopComplete = true
+			serverStopped = nil
+		case <-logstashInstancesStopped:
+			d.log.Debug("logstash instance successfully stopped.")
+			logstashInstanceStopComplete = true
+			logstashInstancesStopped = nil
+		}
+	}
+	t.Stop()
+}
+
+// Cleanup removes the temporary files created by the daemon.
+func (d *Daemon) Cleanup() {
+	err := os.RemoveAll(d.tempdir)
+	if err != nil {
+		d.log.Errorf("Failed to cleanup temporary directory for daemon %q: %v", d.tempdir, err)
+	}
+}
+
+// Shutdown signals the daemon to shutdown.
+func (d *Daemon) Shutdown(ctx context.Context, in *pb.ShutdownRequest) (*pb.ShutdownResponse, error) {
+	select {
+	case d.shutdownSignal <- struct{}{}:
+	default:
+	}
+
+	return &pb.ShutdownResponse{}, nil
+}
+
+func (d *Daemon) isShutdownInProgress() bool {
+	select {
+	case <-d.shutdownInProgress:
+		return true
+	default:
+	}
+	return false
+}
+
+// SetupTest creates a new session, receives the pipeline configuration
+// (zip archive), and prepares the files for the new session.
+func (d *Daemon) SetupTest(ctx context.Context, in *pb.SetupTestRequest) (*pb.SetupTestResponse, error) {
+	if d.isShutdownInProgress() {
+		return nil, errors.New("daemon is shutting down, no new sessions accepted")
+	}
+
+	pipelines, configFiles, err := d.extractZip(in.Pipeline)
+	if err != nil {
+		return nil, err
+	}
+
+	session, err := d.sessionController.Create(pipelines, configFiles)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.SetupTestResponse{
+		SessionID: session.ID(),
+	}, err
+}
+
+func (d *Daemon) extractZip(in []byte) (pipeline.Pipelines, []logstashconfig.File, error) {
+	r, err := zip.NewReader(bytes.NewReader(in), int64(len(in)))
+	if err != nil {
+		return pipeline.Pipelines{}, nil, err
+	}
+
+	pipelines := pipeline.Pipelines{}
+	configFiles := make([]logstashconfig.File, 0, len(r.File))
+	for _, f := range r.File {
+		rc, err := f.Open()
+		if err != nil {
+			return pipeline.Pipelines{}, nil, err
+		}
+
+		switch f.Name {
+		case "pipelines.yml":
+			pipelinesBody, err := ioutil.ReadAll(rc)
+			if err != nil {
+				return pipeline.Pipelines{}, nil, err
+			}
+
+			err = yaml.Unmarshal([]byte(pipelinesBody), &pipelines)
+			if err != nil {
+				return pipeline.Pipelines{}, nil, err
+			}
+		default:
+			body, err := ioutil.ReadAll(rc)
+			if err != nil {
+				return pipeline.Pipelines{}, nil, err
+			}
+			configFile := logstashconfig.File{
+				Name: f.Name,
+				Body: body,
+			}
+			configFiles = append(configFiles, configFile)
+		}
+
+		err = rc.Close()
```

Comment:
This is the only place we're closing `rc`, i.e. not in the error returns on lines 259, 264, and 269.

One way to reduce the duplication when Close() calls are introduced would be to extract the ReadAll() call to outside the switch block.

Reply:
good catch, I cleaned this up.